trinity/protocol/common/managers.py

```diff
@@ -81,13 +82,31 @@ def __init__(
         if timeout is None:
             timeout = self.response_timout
 
-        self._request(request)
-        while self._is_pending():
-            try:
-                yield await self._get_payload(timeout)
-            except TimeoutError:
-                tracker.record_timeout(timeout)
-                raise
+        start_at = time.perf_counter()
+
+        # The _lock ensures that we never have two concurrent requests to a
+        # single peer for a single command pair in flight.
+        try:
+            await self.wait(self._lock.acquire(), timeout=timeout)
+        except TimeoutError:
+            raise AlreadyWaiting(
+                "Timed out waiting for {0} request lock for peer: {1}".format(
+                    self.response_msg_name,
+                    self._peer
+                )
+            )
+
+        try:
+            self._request(request)
+            while self._is_pending():
+                timeout_remaining = max(0, timeout - (time.perf_counter() - start_at))
+                try:
+                    yield await self._get_payload(timeout_remaining)
+                except TimeoutError:
+                    tracker.record_timeout(timeout)
+                    raise
+        finally:
+            self._lock.release()
```

Comment:
Wait, does this `finally` account for both `try` blocks? If not, don't we need to free the lock in another `finally` for the first `try` as well?

Reply:
You mean the inner `try`? Yeah, the exception will bubble up and this outer `finally` will run.

Reply:
Not the inner, the one above. This is where we acquire the lock.

https://github.com/ethereum/py-evm/pull/1232/files/d3a6d95b5ca6663e35e56a36f7b7bdf70a91cb35#diff-31d44fca310c7755ffc8e946f659e7c3R89

And then we raise `AlreadyWaiting` and jump out of this method. So are we releasing the lock with that `finally` at the bottom?

Reply:
Ah, scratch that. We do not acquire the lock there, hence the exception in the first place.